mqtt/mqtt-bridge/src/persist/loader.rs

```diff
@@ -1,48 +1,69 @@
-use std::{pin::Pin, sync::Arc, task::Context, task::Poll, vec::IntoIter};
+#![allow(dead_code)] // TODO remove when ready
+
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
 
 use futures_util::stream::Stream;
 use mqtt3::proto::Publication;
-use parking_lot::{Mutex, MutexGuard};
+use parking_lot::Mutex;
 
-use crate::persist::{memory::waking_map::WakingMap, Key};
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
 
 /// Message loader used to extract elements from bridge persistence
+///
 /// This component is responsible for message extraction from the persistence
 /// It works by grabbing a snapshot of the most important messages from the persistence
 /// Then, will return these elements in order
+///
 /// When the batch is exhausted it will grab a new batch
-pub struct InMemoryMessageLoader {
-    state: Arc<Mutex<WakingMap>>,
-    batch: IntoIter<(Key, Publication)>,
+pub struct MessageLoader<S: StreamWakeableState> {
+    state: Arc<Mutex<S>>,
+    batch: VecDeque<(Key, Publication)>,
     batch_size: usize,
 }
 
-impl InMemoryMessageLoader {
-    pub async fn new(state: Arc<Mutex<WakingMap>>, batch_size: usize) -> Self {
-        let batch: IntoIter<(Key, Publication)> = Vec::new().into_iter();
+impl<S: StreamWakeableState> MessageLoader<S> {
+    pub fn new(state: Arc<Mutex<S>>, batch_size: usize) -> Self {
+        let batch = VecDeque::new();
 
-        InMemoryMessageLoader {
-            state: Arc::clone(&state),
+        Self {
+            state,
             batch,
             batch_size,
         }
     }
+
+    fn next_batch(&mut self) -> Result<VecDeque<(Key, Publication)>, PersistError> {
+        let mut state_lock = self.state.lock();
+        let batch: VecDeque<_> = state_lock
+            .batch(self.batch_size)?
+            .iter()
+            .map(|(key, publication)| (key.clone(), publication.clone()))
```

Comment:
it feels like we have unneccessary cloning here. Lets make `StreamWakeableState::batch` to return VecDeque it use it without cloning?
`iter::collect` can collect to `VecDeque` as well, so minimal code changes